node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -966,8 +942,54 @@ async fn handle_our_view_change(state: &mut State, view: OurView) -> Result<()>
 		state.waiting_collation_fetches.remove(removed);
 	}
 
+	let new_leaves: Vec<_> = view.difference(&state.view).cloned().collect();
 	state.view = view;
 
+	let id = match state.collating_on {
+		Some(id) => id,
+		None => return Ok(()),
+	};
+
+	for relay_parent in new_leaves {
+		tracing::debug!(
+			target: LOG_TARGET,
+			?relay_parent,
+			para_id = ?id,
+			"Processing new relay parent.",
+		);
+
+		// Determine our assigned core.
+		// If it is not scheduled then ignore the relay parent.
+		let (our_core, num_cores) = match determine_core(ctx, id, relay_parent).await? {
+			Some(core) => core,
+			None => continue,
+		};
+
+		// Determine the group on that core.
+		let current_validators =
+			determine_our_validators(ctx, runtime, our_core, num_cores, relay_parent).await?;
+
+		let validators = current_validators.validators;
+		let no_one_is_assigned = validators.is_empty();
+
+		if no_one_is_assigned {
+			continue
+		}
+
+		tracing::debug!(
+			target: LOG_TARGET,
+			?relay_parent,
+			?validators,
+			para_id = ?id,
+			"Connecting to validators.",
+		);
+
+		// Add the current validator group to the reserved peers
+		connect_to_validators(ctx, validators).await;
```

Comment:
That might cause issues. `new_leaves` is an `Option` at the moment, but that does not really fix the issue: We are not only collating on the very latest leaf. If we get a new leaves update for forks, we will only be able to collate with validators on the fork which had the most recent leaf update (Assuming other forks have different block numbers and thus might have other validators assigned.) 

We could  collect validators of all currently active leaves. Which in turn means we would stay connected to validators on stale forks, until it's block height gets finalized - so it is not perfect either.

What can happen with this implementation is, that collators will not be able to collate on the "right" fork at rotation boundaries sometimes. Considering that rotation boundaries are tough anyways as of now, this should actually be fine for the interim solution this is. ... so disregard.

Reply:
This is a valid concern which I also have considered, but as you pointed out this is an interim solution.

Reply:
> new_leaves is an Option at the moment, but that does not really fix the issue: We are not only collating on the very latest leaf

note that we're using OurView instead of ActiveLeavesUpdate, so `new_leaves` can be a few leaves

Reply:
0ad4aeb